packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2725,6 +2743,20 @@ export class ContainerRuntime
 			throw e;
 		}
 	}
+
+	/**
+	 * Process an empty batch, which will execute expected actions while processing even if there are no messages.
+	 * This is a separate function because the processCore function expects at least one message to process.
+	 */
+	private processEmptyBatch(sequenceNumber) {
+		assert(sequenceNumber !== undefined, "sequenceNumber must be defined");
+		this.emit("batchBegin", { sequenceNumber });
+		if (!this.hasPendingMessages()) {
+			this.updateDocumentDirtyState(false);
+		}
+		this.emit("batchEnd", undefined, { sequenceNumber });
+	}
```

Comment:
You also want to update `this._processedClientSequenceNumber` (use `batchStartCsn`) and to call `this.resetConnectCount()` (if `local`).

I think it's fine to skip emitting the "op" event. But you could just as well pass the empty Grouped Batch op.  Probably leave it as-is for now.

Reply:
I think it's fine to skip emitting the "op" event. But you could just as well pass the empty Grouped Batch op.  Probably leave it as-is for now to keep moving, but we can keep it in mind if needed in the future.